[DX-1008] Updated Internal Endpoint middleware (#4046)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/middleware/internal-endpoint-tyk-oas.md

---
title: Using the Internal Endpoint middleware with Tyk OAS APIs
date: 2024-01-26
description: "Using the Internal Endpoint middleware with Tyk OAS APIs"

> tags: ["internal endpoint", "internal", "middleware", "per-endpoint", "Tyk OAS"]

Author: I haven't put "Tyk OAS APIs" in any of the other pages - if you want this added then for consistency this should be in all of the middleware pages (and "Tyk Classic APIs" for all of the Tyk Classic pages). Do you want these labels added to all of the PRs?

---

The [Internal Endpoint]({{< ref "product-stack/tyk-gateway/middleware/internal-endpoint-middleware" >}}) middleware instructs Tyk Gateway not to process external requests to the endpoint (which is a combination of HTTP method and path). Internal requests from other APIs will be processed.

> When working with Tyk OAS APIs, the middleware is configured in the [Tyk OAS API Definition]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc#operation" >}}) either manually within the `.json` file or from the API Designer in the Tyk Dashboard.

Author: ```suggestion
When working with Tyk OAS APIs, the middleware is configured in the [Tyk OAS API Definition]({{< ref "tyk-apis/tyk-gateway-api/oas/x-tyk-oas-doc#operation" >}}). You can do this via the Tyk Dashboard API or in the API Designer.
```

        "info": {
            "name": "example-internal-endpoint",
            "state": {
                "active": true
            }
        },
        "upstream": {
            "url": "http://httpbin.org/"
        },
        "server": {
            "listenPath": {
                "value": "/example-internal-endpoint/",
                "strip": true
            }
        },
        "middleware": {
            "operations": {
                "anythingget": {
                    "internal": {
                        "enabled": true
                    }
                },
                "redirectget": {
                    "urlRewrite": {
                        "enabled": true,
                        "pattern": ".*",
                        "rewriteTo": "tyk://self/anything"
                    }
                }
            }
        }
    }
}
```

In this example, two endpoints have been defined:
 - the internal endpoint middleware has been configured for HTTP `GET` requests to the `/anything` endpoint
 - the [URL rewrite]({{< ref "transform-traffic/url-rewriting" >}}) middleware has been configured for HTTP `GET` requests to the `/redirect` endpoint
 

> Any calls made directly to `GET /example-internal-endpoint/anything` will be rejected, with Tyk returning `HTTP 403 Forbidden`.

Reviewer: ```suggestion
Any calls made directly to `GET /example-internal-endpoint/anything` will be rejected, with Tyk returning `HTTP 403 Forbidden`, since the `/anything` endpoint is internal.
```

Any calls made to `GET /example-internal-endpoint/redirect` will be redirected to `/example-internal-endpoint/anything`. These will be proxied to the upstream because they originate from within Tyk Gateway (i.e. they are internal requests) - so the response from `GET http://httpbin.org/anything` will be returned.

> The configuration above is a complete and valid Tyk OAS API Definition that you can import into Tyk to try out the API-level response header transform.

Reviewer: ```suggestion
The configuration above is a complete and valid Tyk OAS API Definition that you can import into Tyk to try out the API-level internal endpoint middleware.
```

Author: ```suggestion
The configuration above is a complete and valid Tyk OAS API Definition that you can import into Tyk to try out the internal endpoint middleware.
```

## Configuring the middleware in the API Designer
Adding the Internal Endpoint middleware to your API endpoints is easy when using the API Designer in the Tyk Dashboard, simply follow the steps taken in this short video:

>  < placeholder for video >

Reviewer: Remember to add placeholder video when available.

Author: I've added screenshots at this time
